### PR TITLE
Connect topic classification

### PR DESCRIPTION
**Pull request by bdittes**
Adds backend and frontend code to fetch, display and accept suggestions from huridocs/topic-classification

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [x] Code review

### CODE REVIEW

**Comment by RafaPolit** [comment]
Our "Ready for QA" label means that the development is finished and that it is pending a QA, but, should the QA pass, the branch is ready to be merged.  If this is not the case, please remove the 'ready for QA'.  We are reviewing the PR, but..., should we green light it, is it ready to be merged?  

Thanks.

**Comment by bdittes** [comment]
I believe test coverage and code quality are now good enough for this to be merged.

**Comment by bdittes** [comment]
Thanks, this is a very helpful review and I can only say



Also, Sam and I have collected all of these into an 'Uwazi Code Style Guide', please extend:

https://docs.google.com/document/d/1vOabzSemZZNepIRUH-d6dsKTWLoUIqX9hW7nnKQ11Kc/edit#heading=h.a0k912j6j4wr

**Comment by daneryl** [comment]
> Also, Sam and I have collected all of these into an 'Uwazi Code Style Guide', please extend

Great, thanks ! will take a look and add anything that comes to mind.

File: app/react/Thesauris/specs/ThesaurusCockpit.spec.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      };
+    });
+
+    const render = () => {
+      component = shallow(<ThesaurusCockpitBase {...props} />, { context });
+    };
+
+    it('should render a ThesaurusCockpit', () => {
+      render();
+      expect(component).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/react/Thesauris/specs/ThesaurusCockpit.spec.js]: Done.

File: app/react/Thesauris/specs/ThesaurusCockpit.spec.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      thesaurus.property = {
+        content: 'content1',
+        label: 'ThesaurusName',
+        name: 'thesaurus_name',
+      };
+      props = { models, thesaurus: thesauri[0], suggestions };
+      RouteHandler.renderedFromServer = true;
+      dispatchCallsOrder = [];
+      context = {
+        subscribe: jest.fn(), // Find out why this needs to be here.
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/react/Thesauris/specs/ThesaurusCockpit.spec.js]: quite. removed

File: app/react/Thesauris/specs/ThesaurusCockpit.spec.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+describe('ThesaurusCockpit', () => {
+  describe('render', () => {
+    let component;
+    let props;
+    let context;
+    let dispatchCallsOrder = [];
+
+    beforeEach(() => {
+      const thesaurus = thesauri[0];
+      // The render function has already mapped this property back to the thesaurus by now
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/react/Thesauris/specs/ThesaurusCockpit.spec.js]: it was helpful to me as I wrote the test but isn't useful in general. removed.

File: app/api/topicclassification/api.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+export async function checkModelReady(arg: { model: string }) {
+  const tcUrl = new URL('models', tcServer);
+  tcUrl.searchParams.set('model', arg.model);
+  const msTimeout = 1000;
+  if (!(await isReachable(tcUrl.href, { timeout: msTimeout }))) {
+    // TODO: move this backend check to server start-up time, maybe
+    return { models: {}, error: `Topic Classification server is unreachable (${msTimeout})` };
+  }
+  return fetch(tcUrl.href, {
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/api/topicclassification/api.ts]: Done

File: app/api/topicclassification/api.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+export async function processDocument(arg: { seq: string; model: string }) {
+  const url = new URL(tcServer);
+  url.searchParams.set('model', arg.model);
+  const res = await fetch(url.pathname, {
+    method: 'POST',
+    body: JSON.stringify(arg.seq),
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  });
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/api/topicclassification/api.ts]: Done

File: app/api/topicclassification/api.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  })
+    .then(async res => res.json())
+    .catch(err => {
+      return {
+        models: '',
+        error: `Error from topic-classification server: ${err.toString()}`,
+      };
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/api/topicclassification/api.ts]: This is an open point of discussion. Right now, the coupling has been done as a soft-dependency -- if the topic-classification server is unavailable, everything else in uwazi should continue to work. This was my way of noting that it was broken, but not interfering with the rest of the page load.

Open to alternative approaches if you think they ought to be considered.

File: app/api/topicclassification/routes.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+const tcUrlPrefix = '/api/models';
+
+export default (app: Application) => {
+  app.get(
+    tcUrlPrefix,
+    needsAuthorization(),
+    validation.validateRequest(Joi.object().keys({ model: Joi.string() })),
+    (req: Request, res: Response, next: NextFunction) => {
+      if (typeof req.query.model === 'undefined') {
+        getAllModels(req.query.model)
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/api/topicclassification/routes.ts]: very true.

File: app/api/topicclassification/topicClassification.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+export async function classify(e: EntitySchema, thesaurusName: string) {
+  const seq = await extractSequence(e);
+  const modelName = buildModelName(thesaurusName);
+  const results = await processDocument({
+    seq,
+    model: modelName,
+  });
+  return results;
+}
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/api/topicclassification/topicClassification.ts]: removed

File: app/react/Forms/components/MultiSelect.js [hunk 9]
```diff
@@ -255,6 +255,7 @@ export default class MultiSelect extends Component {
 
   render() {
     let { optionsValue, optionsLabel } = this.props;
+    const { thesaurusName } = this.props;
     optionsValue = optionsValue || 'value';
     optionsLabel = optionsLabel || 'label';
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/react/Forms/components/MultiSelect.js]: I don't think that allows mixing 'let' and 'const'
https://stackoverflow.com/questions/47263645/how-to-mix-const-and-let-when-using-object-or-array-destructuring-assignment-in

File: app/react/Forms/components/MultiSuggest.tsx [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+          )))()}
+      </div>
+    );
+  }
+}
+
+function mapStateToProps(state: any, props: MultiSuggestProps) {
+  let { selectModel } = props;
+  if (selectModel && selectModel[0] === '.') {
+    // TODO(bdittes): Correctly inherit parent path.
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/react/Forms/components/MultiSuggest.tsx]: Yes

File: app/react/Review/components/OneUpEntityViewer.js [hunk 11]
```diff
@@ -0,0 +1,509 @@
+/** @format */
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/react/Review/components/OneUpEntityViewer.js]: It's mostly tested through e2e, which is not visible via coverage...

File: app/react/Forms/components/MultiSelect.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (daneryl) [reply app/react/Forms/components/MultiSelect.js]: @bdittes optionsValue and optionsLabel are not actually reassigned and have defaultProps configured as 'value', 'label', reassigning on lines 258 and 259 i think its redundant.

File: app/react/Thesauris/ThesaurusCockpit.tsx [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+export class ThesaurusCockpitBase extends RouteHandler {
+  static genIcons(label: string, actual: number, possible: number) {
+    const icons = [];
+    for (let i = 0; i < possible; i += 1) {
+      let iconClass: any;
+      if (i < actual) {
+        iconClass = 'circle';
+      } else {
+        iconClass = ['far', 'circle'];
+      }
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/react/Thesauris/ThesaurusCockpit.tsx]: Done

File: app/react/Settings/Settings.js [hunk 13]
```diff
@@ -1,30 +1,97 @@
...
+          // may be counted more than once.
+          // TODO: Make document counts unique.
+          if (!thesaurus.hasOwnProperty('suggestions')) {
+            thesaurus.suggestions = 0;
+          }
+          thesaurus.suggestions += soFar;
+        }
+      });
+    });
+
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/react/Settings/Settings.js]: (I may have misunderstood something here, but...) My impression was that all of Settings requestState is currently (monolithically) handled in Settings.js, probably to facilitate and speed up moving between various settings consoles. E.g. thesauriList.js is a component, not a routehandler.

I expected that there was some wisdom in this, and that there would be value in keeping ThesauriList consistent with the others. If you disagree, I'm happy to refactor this and make ThesauriList a routehandler that fetches its own state. At a high level, I think that's more elegant anyway (and doesn't clutter the state unnecessarily when this console isn't clicked.

File: app/api/topicclassification/topicClassification.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+export async function getAllModels(thesauri: string[]) {
+  const results =
+    thesauri !== undefined
+      ? await Promise.all(
+          thesauri.map(async thesaurus => checkModelReady({ model: buildModelName(thesaurus) }))
+        )
+      : await getModels();
+  return results;
+}
```
[the remark this answers is not in the document]
Comment (samschaevitz) [reply app/api/topicclassification/topicClassification.ts]: Indeed. I'm going to clean this API up and add tests for it...

File: nightmare/paths/review.spec.js [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        .waitToClick(selectors.review.toggleFullEditButton)
+        .waitToClick(selectors.review.toggleFullEditButton)
+        .waitToClick(selectors.review.toggleFullEditButton)
+        .waitToClick(selectors.review.toggleFullEditButton)
+        .waitToClick(selectors.review.toggleFullEditButton);
+      expect(await nightmare.isVisible(selectors.review.titleEditBox)).toBe(true);
+    });
+
+    it('should allow accepting a suggestion', async () => {
+      await nightmare.waitToClick(selectors.review.firstSuggestion).wait(100);
```
[the remark this answers is not in the document]
Comment (bdittes) [reply nightmare/paths/review.spec.js]: Fun fact: waitToClick includes a fixed wait(300).
But I did remove them as much as possible.

File: app/react/Uploads/UploadsRoute.js [hunk 16]
```diff
@@ -105,22 +104,27 @@ export default class Uploads extends RouteHandler {
 
   render() {
     const query = rison.decode(this.props.location.query.q || '()');
+    const viewOnly = this.props.location.query.view === 'nosearch';
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/react/Uploads/UploadsRoute.js]: Maybe eventually, for now it's a page that allows publishing docs without search or filters, so I think it's close enough to the existing /uploads.

File: app/react/Review/specs/OneUpReview.spec.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
+/** @format */
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/react/Review/specs/OneUpReview.spec.js]: Please advise on how to test requestParams()

File: nightmare/paths/review.spec.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (daneryl) [reply nightmare/paths/review.spec.js]: thx, created an issue to remove the fixed wait 300

**Comment by RafaPolit** [comment]
We have reviewed this with @daneryl .

The code is good, we are happy with all the changes that have happened.  If you could fix the functions reported by the codeclimate as having too deep of a complexity, that would be great.

As for functionality, here is what we have found, which are more questions at this point but some are suggestions:

- since the entity title cannot use the top bar (where now a lot of the functionality is located at), moving the title to display with the 'entity metadata' has not been fully solved design-wise: title is bolded like the labels for properties (not values), the type of template is floating at the right.  This should be evolved another iteration to display: "title", "template" labels or other solution.

- If I 'manually' return to the original entity's state (by selecting and unselecting suggested values), I'm still presented with a "discard changes" even though I have already discarded them manually. Shouldn't this be more of a comparisson flow, where if you have the exact original entity values, you don't have the option to discard? (minor detail)

As for the flow:
Fixtures only encompass a single document, so it is hard to test the flow, but:
- The "next" arrow is enabled, even if I only have one document available, we assume that reaching the last document has this same error
- Since there is only one single document, it's hard to know if saving a doc correctly takes only that off the 'to review' list leaving all others or if the list is being cleaned even after a single review.
- There is no way for us to test the "confidence" stop sign and that it grows correctly
- What happens if two properties are up for review from different thesauri? Do they both show on the right panel? Or am I only editting "that" particular thesaurus

We feel that, at the very least, the e2e should test two documents to include the 'next' or 'save and next' flow, which is an fundamental part of the process.

**Comment by bdittes** [comment]
> * since the entity title cannot use the top bar (where now a lot of the functionality is located at), moving the title to display with the 'entity metadata' has not been fully solved design-wise: title is bolded like the labels for properties (not values), the type of template is floating at the right.  This should be evolved another iteration to display: "title", "template" labels or other solution.

I'm just a lowly engineer implementing UX designs. They even went through user testing and I didn't hear about any confusion... But I have made the title a bit stronger and move the template closer:
image

> * If I 'manually' return to the original entity's state (by selecting and unselecting suggested values), I'm still presented with a "discard changes" even though I have already discarded them manually. Shouldn't this be more of a comparisson flow, where if you have the exact original entity values, you don't have the option to discard? (minor detail)

I agree that would be even nicer, but it's also much harder to implement (currently I'm just using redux entityForm.pristine). I also didn't hear about user confusion about this, so I'd vote to keep it as is.

> As for the flow:
> Fixtures only encompass a single document, so it is hard to test the flow, but:
> 
> * The "next" arrow is enabled, even if I only have one document available, we assume that reaching the last document has this same error

Done.

> * Since there is only one single document, it's hard to know if saving a doc correctly takes only that off the 'to review' list leaving all others or if the list is being cleaned even after a single review.

Mostly for performance reasons, navigation between documents only does minimal backend calls. All sharedIds to be reviewed are loaded initially, so navigating just loads that one doc. See the updated e2e test review.spec.js.

> * There is no way for us to test the "confidence" stop sign and that it grows correctly

I don't fully understand this one. Do you mean how does it look if a suggestion is really long? That looks like this:
image


> * What happens if two properties are up for review from different thesauri? Do they both show on the right panel? Or am I only editting "that" particular thesaurus

Per the [design](https://www.figma.com/file/C6GTub8a3jLamXJMiXjfpS/1UP?node-id=708%3A3214), the right side shows all 'ml-enabled' thesauri. 

> 
> We feel that, at the very least, the e2e should test two documents to include the 'next' or 'save and next' flow, which is an fundamental part of the process.

Done.

**Comment by daneryl** [comment]
@bdittes still i think some lint and codeclimate fixes are missing, but i think this can be merged for now and we can create an issue to fix those as tech debt ? what you think @RafaPolit ?

**Comment by bdittes** [comment]
> @bdittes still i think some lint and codeclimate fixes are missing, but i think this can be merged for now and we can create an issue to fix those as tech debt ? what you think @RafaPolit ?

I went over the lint warning and codeclimate - due to some eslint changes, this PR touches a lot of files, which re-issues these in code we didn't really touch.

**Pull request merged**